src/bin/nydus-image/deduplicate.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 Ant Group. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Deduplicate for Chunk
+
+use std::path::Path;
+use std::sync::Arc;
+
+use anyhow::{Context, Result};
+use nydus_api::ConfigV2;
+use nydus_rafs::builder::Tree;
+use nydus_rafs::metadata::RafsSuper;
+use nydus_storage::device::BlobInfo;
+use nydus_storage::meta::format_blob_features;
+use rusqlite::{params, Connection};
+pub struct Deduplicate {
+    sb: RafsSuper,
+}
+
+impl Deduplicate {
+    pub fn new(bootstrap_path: &Path, config: Arc<ConfigV2>) -> Result<Self> {
+        let (sb, _) = RafsSuper::load_from_file(bootstrap_path, config, false)?;
+        Ok(Self { sb })
+    }
+
+    /// save metadata to database: chunk and blob info
+    pub fn save_metadata_to_database(&mut self) -> Result<Vec<Arc<BlobInfo>>> {
+        let err = "failed to load bootstrap for deduplicate";
+        let tree = Tree::from_bootstrap(&self.sb, &mut ()).context(err)?;
+
+        // Connect to the database
+        let database_file: &str = "metadata.db";
```

Comment:
We need to support various database protocols (not only for sqlite), so I think a abstract `Database` trait should be introduced, like:

```
trait Database {
    fn insert_chunk(&mut self, chunk: &Chunk) -> Result<()>;
    fn insert_blob(&mut self, blob: &Blob) -> Result<()>;
    fn get_chunk(&self) -> Result<Chunk>;
    fn get_blob(&self) -> Result<Blob>;
}
```

Reply:
And please also add some unit test cases.

Reply:
Ok.